library/src/cache.rs

```diff
@@ -179,13 +182,44 @@ impl UpdaterState {
         None
     }
 
+    fn validate(&mut self) -> anyhow::Result<()> {
+        // iterate through all slots:
+        // Make sure they're still valid.
+        // If not, remove them.
+        let slot_count = self.slots.len();
+        let mut needs_save = false;
+        // Iterate backwards so we can remove slots.
+        for i in (0..slot_count).rev() {
+            let slot = &self.slots[i];
+            if !self.validate_slot(slot) {
+                warn!("Slot {} is invalid, clearing.", i);
+                self.clear_slot(i)?;
+                needs_save = true;
+            }
+        }
+        if needs_save {
+            self.save()?;
+        }
+        Ok(())
+    }
+
     fn validate_slot(&self, slot: &Slot) -> bool {
         // Check if the patch is known bad.
-        if self.is_known_bad_patch(&slot.to_patch_info()) {
+        if self.is_known_bad_patch(&PatchInfo {
+            path: String::new(),
+            number: slot.patch_number,
+        }) {
```

Comment:
I'd add a fixme here to update is_known_bad_patch to only require a patch number